etc/testing/circle/workloads/go-test-results/collector/main.go

```diff
@@ -54,52 +51,144 @@ func main() {
-	}
-	jobName := sanitizeName(os.Getenv("CIRCLE_JOB"))
-	if len(jobName) == 0 {
-		logger.Fatal("CIRCLE_JOB needs to be populated to upload test results.")
+	if _, err := os.Stat(resultsFolder); os.IsNotExist(err) {
+		logger.Fatalf("The test result folder at %v does not exist. Exiting early.", resultsFolder)
 	}
-
+	// connect and authenticate to pachyderm
 	pachClient, err := client.NewFromURIContext(context.Background(), pachdAddress)
 	if err != nil {
 		logger.Fatalf("Provisioning pach client: %v", err)
 	}
 	pachClient.SetAuthToken(robotToken)
-
+	// retry in case the parent commit is still in progress from another pipeline
+	var commit *pfs.Commit
+	for i := 0; i < commitMaxRetries; i++ {
+		commit, err = pachClient.StartProjectCommit(projectName, repoName, "master")
+		if err != nil {
+			logger.Printf("Unable to start commit due to error. Retrying. Error: %v", err)
+			time.Sleep(commitRetryBackoff)
+		} else {
+			break
+		}
+	}
+	if err != nil {
+		logger.Fatalf("Starting commit: %v", err)
+	}
+	defer func() {
+		if err = pachClient.FinishProjectCommit(projectName, repoName, "master", commit.GetID()); err != nil {
+			logger.Fatalf("Finishing commit: %v", err)
+		}
+	}()
+	// upload general job information
+	jobInfo, err := findJobInfoFromCI()
+	if err != nil {
+		logger.Fatalf("Collecting job info: %v", err)
+	}
+	basePath := findBasePath(jobInfo)
+	if err = uploadJobInfo(basePath, jobInfo, pachClient, commit); err != nil {
+		logger.Fatalf("Uploading job info: %v", err)
+	}
+	// upload individual test  results
 	eg, _ := errgroup.WithContext(pachClient.Ctx())
 	err = filepath.WalkDir(resultsFolder, func(path string, d fs.DirEntry, err error) error {
 		if !d.IsDir() && strings.HasSuffix(d.Name(), fileSuffix) {
 			eg.Go(func() error {
-				file, err := os.Open(path)
-				if err != nil {
-					return errors.Wrapf(err, "opening file %v", d.Name())
-				}
-				defer file.Close()
-
-				destPath := findDestinationPath(path, branchFolderName, jobName, resultsFolder)
-				commit := client.NewProjectCommit(projectName, repoName, "master", "")
-				err = pachClient.PutFile(commit, destPath, file)
-				if err != nil {
-					return errors.Wrapf(err, "putting file: %v to commit %v", destPath, commit.GetID())
-				}
-				logger.Printf("Succeeded uploading %s \n", path)
-				return nil
+				return uploadTestResult(path, d, basePath, resultsFolder, pachClient, commit)
 			})
 		}
 		return nil
 	})
 	if err != nil {
-		logger.Fatalf("Failed walking file paths: %v", err)
+		logger.Fatalf("Walking file paths: %v", err)
 	}
 	if goRoutineErrs := eg.Wait(); goRoutineErrs != nil {
-		logger.Fatalf("Failed creating files: %v", err)
+		logger.Fatalf("Putting files: %v", err)
 	}
+
 	logger.Println("Successfully uploaded files.")
 }
+
+func sanitizeName(s string) string {
+	ret := []byte(strings.ReplaceAll(s, " ", "_"))
+	invalidCharacters := regexp.MustCompile("[^a-zA-Z0-9-_]+")
```

Comment:
I'd lift regexp.MustCompile out to the package scope; `var invalidNameCharacters = regexp.MustCompile(...)`.